Improved and clearer Java installation tutorial with screenshots (#133)
MinecraftHopper/MinecraftHopper: _help/installing-java/index.markdown

desc: "How to install Java"
---
# What is Java<small> - How to install Java</small>

The Java Runtime Environment (JRE) is software that allows for programs written in the Java programming language to run on your computer. "Minecraft Java Edition" is named as such because it is written in Java. Java comes in many shapes and forms, which may make it confusing exactly which version of it you should have installed, which is what this should hopefully explain.

We cannot use **Oracle Java** for modern versions of Java, since they have commercial usage restrictions which mean some users (particularly, those making money off of Minecraft) would have to pay to use it.

Fortunately, we can use **OpenJDK**, which is identical to Oracle Java but does not have the same restrictions. For this tutorial we will be using a version of OpenJDK built by [Adoptium](https://adoptium.net).

Quick links:

> - [Do I need Java? Which version do I need?](#which-java)

Reviewer: You need to change this to reflect the new order of the sections.

![Terminal with install command typed in](/static/images/help/installing-java/linux/install-java-2.jpg)

![Terminal with install command executed](/static/images/help/installing-java/linux/install-java-3.jpg)

If you install multiple versions of Java, you may occasionally need to change which one is used by default. To do so:

##### Ubuntu/Debian/Pop!\_OS/Linux Mint

To show available Java versions:
```
update-java-alternatives --list
```
To switch to use Java 17 by default:
```
sudo update-java-alternatives --set java-1.17.0-openjdk-amd64
```

##### Fedora/RHEL/CentOS/RockyLinux/Amazon Linux

```
sudo alternatives --config java
```

Follow the interactive prompt

##### Archlinux/EndeavorOS/Manjaro

To show available Java versions:
```
archlinux-java status
```
To switch to use Java 17 by default:
```
sudo archlinux-java set java-17-openjdk
```

![Terminal with Java alternatives command executed](/static/images/help/installing-java/linux/install-java-4.jpg)

" | markdownify }} </details>

## Next steps

**Additional info:** [Changing the default Java version used by the launcher](/help/guides/changing-java-version/)

> **Additional info:** [Verifying if Java is installed properly](#verifying)

Reviewer: Move this above the Changing launcher. They should verify they have installed Java correctly before trying to use it for anything.

Once you have installed the proper Java version, you should configure this in your launcher and give the game a go! Depending on what you are trying to do, you may need to manually navigate to the path of Adoptium OpenJDK.

Here are some examples of what this path may look like on **Windows**:

```
C:\Program Files\Eclipse Adoptium\jdk-17.0.2.8-hotspot\bin\javaw.exe
C:\Program Files\Eclipse Adoptium\jdk-11.0.14.1.1-hotspot\bin\javaw.exe
C:\Program Files\Eclipse Adoptium\jdk-8.0.322.6-hotspot\bin\javaw.exe
```

Here are some examples of what this path may look like on **macOS**:

```
/Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home/bin/java
/Library/Java/JavaVirtualMachines/temurin-11.jdk/Contents/Home/bin/java
/Library/Java/JavaVirtualMachines/temurin-8.jdk/Contents/Home/bin/java
```

Here are some examples of what this path may look like on **Linux**:

```
/usr/lib/jvm/java-17-openjdk/bin/java
/usr/lib/jvm/java-11-openjdk/bin/java
/usr/lib/jvm/java-8-openjdk/bin/java
/usr/lib/jvm/java-17-openjdk-amd64/jre/bin/java
/usr/lib/jvm/java-11-openjdk-amd64/jre/bin/java
/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java
```

> ## Verifying if Java is installed {#verifying}

Reviewer: Move this above the tutorial section.

Fortunately, we can use **OpenJDK**, which is identical to Oracle Java but does not have the same restrictions. For this tutorial we will be using a version of OpenJDK built by [Adoptium](https://adoptium.net).

Quick links:

- [Downloading and installing Java](#downloading-java)
- [Changing Java runtime settings in third-party launchers (MultiMC)](#multimc)
- [Check what Java version you already have installed](#verifying)
- [Symptoms of incorrect Java version](#symptoms)

## Do I need Java? Which version do I need?

The Minecraft launcher includes the version(s) of Java required to run the game, and automatically keeps them up to date. This includes the Windows 7 and Windows 8.1 version of the launcher.

If you are using the official launcher, you normally do not have to manually install Java unless you need it for applications other than Minecraft. If you are not are not sure if you are using the correct Minecraft launcher or are getting "This application requires a Java Runtime Environment" when launching Minecraft, then you can download the Minecraft Launcher from [https://minecraft.net/download](https://minecraft.net/download).

> The below compatibility chart was last updated 2022-02-20.

Reviewer: Get rid of the EOL links on the table, they're confusing for the average person. I'd recommend adding something above the update notice instead, like so:
```suggestion
**Note:** Java versions 6, 8, and 16 no longer receive security updates, and are therefore more risky to use. You should use Java 11 or 17 if you can, as they will receive security updates for years to come.

The below compatibility chart was last updated 2022-02-20.
```
Java 8 EOLs on March 31st, so I also added that to the note.

| Version of Minecraft | Recommended Java version | Minimum Java version |
|:-:|-|-|
| Future versions not listed | Latest you can find | (unknown) |
| **1.18** | **Java 17** | 17 |
| **1.17** | **Java 17** | 16 [(do not use!)](https://endoflife.date/java) |
| Some Spigot plugins before 1.17 | | 11 |
| **1.12** to **1.16** (inclusive) | **Java 8** | 8 |
| **Classic** to **1.11** (inclusive) | **Java 8** | 6 [(do not use!)](https://endoflife.date/java) |

### Downloading Java {#downloading-java}

Visit the website for [Adoptium Temurin OpenJDK](https://adoptium.net)

> Choose the appropriate version, and click **Latest release** to start the download. Note that Temurin versions are the same as Java versions, so "Temurin 17" means "Temurin Java 17". See the above chart to determine which version you need.

Reviewer: You need to expand this section to explain how to get older versions of Java/OpenJDK.

### Linux installation instructions {#install-linux}

<details><summary>Click for instructions</summary> {{ "

Open a terminal

![KDE menu with terminal selected](/static/images/help/installing-java/linux/install-java-1.jpg)

Install Java with your package manager.

##### Ubuntu / Debian / Pop!\_OS / Linux Mint

To install Java 17, 11, 8 respectively, choose one of these commands:
```
sudo apt install openjdk-17-jdk
sudo apt install openjdk-11-jdk
sudo apt install openjdk-8-jdk
```

There are also PPAs available for Adoptium.

##### Fedora

To install Java 17, 11, 8 respectively, choose one of these commands:
```
sudo dnf install java-latest-openjdk.x86_64
sudo dnf install java-11-openjdk.x86_64
sudo dnf install java-1.8.0-openjdk.x86_64
```

##### Archlinux / EndeavorOS / Manjaro

To install Java 17, 11, 8 respectively, choose one of these commands:
```
sudo pacman -S jdk-openjdk
sudo pacman -S jdk11-openjdk
sudo pacman -S jdk8-openjdk
```

##### RHEL / CentOS / RockyLinux / Amazon Linux

To install Java 17, 11, 8 respectively, choose one of these commands:
```
sudo yum install java-17-openjdk
sudo yum install java-11-openjdk
sudo yum install java-1.8.0-openjdk
```

> If your distro does not support Java, the Microsoft OpenJDK or Adoptium downloads can be used on all distros.

Author: Distributions and package managers are roughly the same thing, but yeah that might be clearer

> If your distro does not support Java, the Microsoft OpenJDK or Adoptium downloads can be used on all distros.

Reviewer: Aren't distributions different versions, like Ubuntu, Fedora, Debian etc.

![Terminal with install command typed in](/static/images/help/installing-java/linux/install-java-2.jpg)

![Terminal with install command executed](/static/images/help/installing-java/linux/install-java-3.jpg)

You may need to adjust which Java version is chosen. To do so:

##### Ubuntu / Debian / Pop!\_OS / Linux Mint

To show available Java versions:
```
update-java-alternatives --list
```
To switch to use Java 17 by default:
```
sudo update-java-alternatives --set java-1.17.0-openjdk-amd64
```

##### Fedora / RHEL / CentOS / RockyLinux / Amazon Linux

```
sudo alternatives --config java
```

Follow the interactive prompt

##### Archlinux / EndeavorOS / Manjaro

To show available Java versions:
```
archlinux-java status
```
To switch to use Java 17 by default:
```
sudo archlinux-java set java-17-openjdk
```

![Terminal with Java alternatives command executed](/static/images/help/installing-java/linux/install-java-4.jpg)

" | markdownify }} </details>

## Next steps

> Once you have installed the proper Java version, you should configure this in your launcher and give the game a go! Depending on your launcher you may need to manually navigate to the path of Adoptium OpenJDK.

Reviewer: We actually have an article on this, https://minecrafthopper.net/help/guides/changing-java-version/

It's very, very outdated though, so I will work on updating it myself, and it can be merged before this article.

We don't need to deal with 3rd-party launchers either, you can get rid of this entire section.

Launch MultiMC, and click either **Edit Instance** or **Settings**.

**Important:** Use `Settings` only if you wish to switch *all* your instances to a specific Java version. Upgrading to Java 17 will break any pre-1.17 Forge instances!

![Settings button in MultiMC](/static/images/help/installing-java/multimc/java-in-multimc-1.jpg)

Click **Settings**, go under the **Java** tab, check **✅ Java Installation**, then click **Browse...**

![Java settings in MultiMC](/static/images/help/installing-java/multimc/java-in-multimc-2.jpg)

You may be able to auto-detect Java installations by clicking **Auto-detect...**.

Otherwise, locate `javaw.exe`:

![Browsing for Java executable](/static/images/help/installing-java/multimc/java-in-multimc-3.jpg)

Run the sanity check to ensure you did everything correctly.

![Java sanity check in MultiMC](/static/images/help/installing-java/multimc/java-in-multimc-4.jpg)

" | markdownify }} </details>

## Verifying if Java is installed {#verifying}

> You can check which version of Java may be already installed on your computer by following these instructions.

Reviewer: ```suggestion
You can check if a version of Java is already installed on your computer by following these instructions.
```

Exception in thread "main" java.lang.UnsupportedClassVersionError: net/minecraft/bundler/Main has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 52.0

Exception in thread "main" java.lang.UnsupportedClassVersionError: net/minecraft/bundler/Main has been compiled by a more recent version of the Java Runtime (class file version 60.0), this version of the Java Runtime only recognizes class file versions up to 55.0

Exception in thread "main" java.lang.UnsupportedClassVersionError: net/minecraft/bundler/Main has been compiled by a more recent version of the Java Runtime (class file version 60.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```
```
Your Java Runtime is incompatible. Please edit your installation to use the bundled Java Runtime.
Name: Java version mismatch
```
```

> (When attempting to launch with MultiMC)

Reviewer: We're not dealing with 3rd-party launchers or server software, all of this can go.

Forge can stay, I think, because of its ubiquitousness with normal people (as opposed to enthusiasts or server admins).

Author: Hmm, I'm disappointed this won't be considered in-scope since excluding it limits the usefulness of the article, and excluding the symptoms limits search engine visibility of the article. Most people are installing Java manually because they're trying to get third-party tools or server software working, since if they were using vanilla it would already be handled

Reviewer: I think that, first and foremost, this site is a tool for MCCS. In general, we only support official things, and while I agree that including additional information about 3rd-party launchers would increase its usefulness outside of supporting official things, it would just increase confusion for users not trying to use those tools, and therefore make it less useful for us specifically.

Reviewer: On top of this, we shouldn't be "supporting" (for want of a better term) 3rd-party tools on the site that we won't support on Discord, that's inconsistent.

Minecraft 1.18 Pre Release 2 and above require the use of Java 17
```
```
(When attempting to launch with MultiMC)
Minecraft 21w19a and above require the use of Java 16
```
```
(When attempting to launch PaperMC or Purpur)
Minecraft 1.18 requires running the server with Java 17 or above. Download Java 17 (or above) from https://adoptium.net/
```

If you see any of the above messages, you need to update to Java 17.

*Note: MultiMC, PaperMC, and Purpur are not officially supported by Minecraft*

```
(When attempting to launch Forge)
java.lang.ClassCastException: class jdk.internal.loader.ClassLoaders$AppClassLoader cannot be cast to class java.net.URLClassLoader (jdk.internal.loader.ClassLoaders$AppClassLoader and java.net.URLClassLoader are in module java.base of loader 'bootstrap')
```

If you see the above message, you need to *downgrade* to Java 8, because old Forge does not support Java 11+.

*Note: Forge is not officially supported by Minecraft*

> ```

Reviewer: All of this can go, non-Minecraft things (even if they're Minecraft related) are out of scope for this article.

Thankfully, **Microsoft OpenJDK** is an identical product without any restrictions, and is what Minecraft uses by default. These tutorials will cover how to install it, as well as an alternative for older Java versions.

**Quick links:** 
- [Installing Java on Windows](#install-windows)
- [Installing Java on macOS](#install-macos)
- [Installing Java on Linux](#install-linux)
- [Check if Java is installed](#verifying)

## Do I need Java? Which version do I need?

The Minecraft launchers for Windows, macOS, and Linux ship with a builtin Java that the launcher automatically keeps up to date. This includes the Windows 7 and Windows 8.1 version of the launcher.

If you are using the official launcher, normally you do not have to install Java on your computer unless you need it for applications other than Minecraft. If you are not are not sure if you are using the new Minecraft launcher or are getting "This application requires a Java Runtime Environment" when launching Minecraft, then you can download the Minecraft Launcher from [https://minecraft.net/download](https://minecraft.net/download).

> **TL;DR: Most people running vanilla Minecraft with the default launcher will not need this tutorial!**

Author: Some may arrive at this page via search engines &mdash; how about

> TL;DR: Most people running vanilla Minecraft with the default launcher will not need this tutorial, but if it has been suggested to you as a workaround, then continue.

| **1.12** to **1.16** (inclusive) | **Java 8** | 8 |
| **Classic** to **1.11** (inclusive) | **Java 8** | 7 [(do not use!)](https://endoflife.date/java) |

**Scroll down for complete instructions on how to install Java on Windows, macOS, Linux.**

## Collecting information - 32 bit, 64 bit, or ARM?

You may need to get a 32 bit x86 or possibly even ARM version of Java, depending on your computer.

On Windows, open a terminal or command prompt and run `wmic OS get OSArchitecture`.

![Screenshot of Windows Terminal with the above command run](/static/images/help/installing-java/verifying/verifying-architecture-windows.jpg)

On macOS, under the Apple menu, select `About This Mac`. If "Chip" contains "Intel", you are on x64. If it contains "Apple", you are on ARM.

![Screenshot of "About This Mac" with "Chip" highlighted](/static/images/help/installing-java/verifying/verifying-architecture-macos.png)

On Linux, your package manager will choose the appropriate version automatically. But if you are going with a third party Java distribution on Linux (such as Microsoft, Adoptium, or Azul) you can check manually by running `uname -m` in a terminal.

> ### I am on Windows and I need latest Java (Version 17) {#install-windows}

Author: The original reason for the split is that I'm recommending Microsoft OpenJDK if you just need java 17, 64 bit, (the theory being that this is least chance of unexpected problems since it is an identical binary to what's used in Minecraft &mdash; has the same hash) and Adoptium if you need something older or 32 bit. This is not a great experience, because the Microsoft website is slightly harder to use because it does not auto detect platform for macOS, and of course *only* provides x64 and ARM.

If this article is to only recommend Adoptium, then yeah, this split is not necessary. I'll reorganize it accordingly.

> ### I am on Windows and I need latest Java (Version 17) {#install-windows}

Author: Done.
